### ISSUE TITLE
Need help building for AWS EC2 instances with ARM CPU and T4G GPU

### ISSUE THREAD

**Issue by usamahjundia-gdp** [microsoft/onnxruntime#10612]
Hi, i think this question does not fall under bugs report / feature request nor security vuln, so i made a blank issue.

I am trying to compile onnxruntime 1.10 with CUDA EP on AWS EC2 G5g.xlarge with an ARM CPU and the NVIDIA T4G GPU. Currently, this is the steps i take in doing such (some parts might be irrelevant to onnxruntime but please bear with me here, its for the sake of completeness)

1. apt update and upgrade
2. Install CUDA 11.4
3. Install CUDNN 8.2.4
4. Install python dev libraries (i use python 3.8)
4. Clone the onnxruntime repo and checkout to 1.10  commit/tag
5. create a python environment (this is not mentioned in the docs / build guide but it doesnt work without this) and install numpy and wheel
6. Build using the following command
`./build.sh --config Release --update --build --build_wheel --use_cuda --cuda_home /usr/local/cuda --cudnn_home /usr/local/cuda `
7. Afterwards, i just create another environment for my project and install the wheel file as normal

I expected a smooth run but of course it wont be that smooth, the build is succesful, i can import the library and load a model just fine, but upon inferencing on CUDA EP, i received this error, verbatim

`onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Non-zero status code returned while running Slice node. Name:'Slice_34' Status Message: CUDA error cudaErrorNoKernelImageForDevice:no kernel image is available for execution on the device`

I am using yolov5s model converted to onnx, and i have already tested running the exact same model on another machine (but that time was using prebuilt wheels from pip since its available) on CUDA EP so it shouldn't be due to the op not being supported by onnx.

I explored github issues in hope it would help and things lead to another and lead me to this piece of code here on the cmakelist.txt:

https://github.com/microsoft/onnxruntime/blob/v1.10.0/cmake/CMakeLists.txt#L1586-L1609

I noticed that T4 has 7.5 Compute capability, but judging from the cmakelists for ARM CPU, the max is 7.2

What am i supposed to do with this file? Am i supposed to just pick one and comment the rest?

So sorry if this is a stupid question, this is my first time building onnxruntime from scratch and consulting the docs does not help me

**Comment by usamahjundia-gdp** [comment]
Ahaha it's actually pretty dumb of an issue

the reason this didnt work was because in the cmakelist on the mentioned line, you can see the if branch handling the arm-based architecture only supports up to compute capability 7.2, whilst my machine has ARM CPU + T4G GPU, so just adding flags for cc 7.5 resolves the issue. I think most people building CUDA EP in ARM are using jetson, which is why the flags are only set up for known jetson gpus.

**TL;DR** 

**If you are building for CUDA EP from source for an ARM Machine and you received the same error (might be different OP), look up the compute capability of your machine's GPU and add this line**

`set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -gencode=arch=compute_<YOUR DEVICE CC HERE>,code=sm_<YOUR DEVICE CC HERE>")`

**Under this if construct**

```
if (CMAKE_LIBRARY_ARCHITECTURE STREQUAL "aarch64-linux-gnu")
      # Support for Jetson/Tegra ARM devices
      set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -gencode=arch=compute_53,code=sm_53") # TX1, Nano
      set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -gencode=arch=compute_62,code=sm_62") # TX2
      set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -gencode=arch=compute_72,code=sm_72") # AGX Xavier, NX Xavier
```

**For example if your GPU has compute capability 7.5, the line you add would be**

`set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -gencode=arch=compute_75,code=sm_75")`

**Comment by joaocc** [comment]
Hi. Could you please confirm if ONNXRuntime already supports these instances, in this case, in C# builds? Thanks

**Comment by usamahjundia-gdp** [comment]
> Hi. Could you please confirm if ONNXRuntime already supports these instances, in this case, in C# builds? Thanks

I didn't use C# so i can't say, you may want to check if onnxruntime has prebuilt libraries for C# on ARM64 architecture although i am not really familiar with Cpp. What does the combination of environments you can choose on the onnxruntime getting started menu tell you?

**Comment by joaocc** [comment]
It looks like there is support for C# on Arm64.
<img width="1249" alt="image" src="https://user-images.githubusercontent.com/2837860/159469849-a883cfbe-8474-46c0-ab8a-75ca6b1a095c.png">

**Comment by joaocc** [comment]
Hi [usamahjundia-gdp](https://github.com/usamahjundia-gdp). Do you know if there is anyone working on supporting this scenario? Thx

**Comment by usamahjundia-gdp** [comment]
> Hi [usamahjundia-gdp](https://github.com/usamahjundia-gdp). Do you know if there is anyone working on supporting this scenario? Thx

Hi, not anyone i am currently aware of, i am afraid.
